src/features/web-compat.js

```diff
@@ -11,7 +11,51 @@ function windowSizingFix () {
     window.outerWidth = window.innerWidth
 }
 
+const MSG_WEB_SHARE = 'webShare'
+
+function canShare (data) {
+    if (typeof data !== 'object') return false
+    if (!('url' in data) && !('title' in data) && !('text' in data)) return false // At least one of these is required
+    if ('files' in data) return false // Not supported at the moment
+    if ('title' in data && typeof data.title !== 'string') return false
+    if ('text' in data && typeof data.text !== 'string') return false
+    if ('url' in data) {
+        if (typeof data.url !== 'string') return false
+        try {
+            const url = new URL(data.url, location.href)
+            if (url.protocol !== 'http:' && url.protocol !== 'https:') return false
+        } catch (err) {
+            return false
+        }
+    }
+    if (window !== window.top) return false // Not supported in iframes
+    return true
+}
+
+/** @returns {ShareRequestData} */
+function cleanShareData (data) {
+    /** @type {ShareRequestData} */
+    const dataToSend = {}
+    for (const key of ['title', 'text', 'url']) {
+        if (key in data) dataToSend[key] = data[key]
+    }
+    if ('url' in data) dataToSend.url = (new URL(data.url, location.href)).href // clean url
+    if ('url' in dataToSend && 'text' in dataToSend) {
+        // combine everything into text
+        dataToSend.text = `${dataToSend.text} ${dataToSend.url}`
+        delete dataToSend.url
+    }
+    if (!('url' in dataToSend) && !('text' in dataToSend)) {
+        // if there's only title, create a dummy empty text
+        dataToSend.text = ''
+    }
+    return dataToSend
+}
```

Comment:
Take it or leave it, but this is a little simpler to read (personally):
```suggestion
/**
 * @returns {ShareRequestData}
 */
function cleanShareData (data) {
    const { title, text, url } = data;
    if (url) url = (new URL(url, location.href)).href // clean url
    if (url && text) {
        // combine everything into text
        text = `${text} ${url}`
        url = undefined
    }
    if (!url && !text) {
        // if there's only title, create a dummy empty text
        text = ''
    }
    return { title, text, url }
}
```

Reply:
There are some caveats with empty strings and undefined values, but I'll take some of your ideas here :)

Reply:
If url / text is empty do we really want to do anything special? Maybe the final fallback to `text=''` is enough?

Reply:
There are subtle differences. For example, if url is an empty string, the effective value is the current page address.
The internal convention with Android we came up with is to pass a cleaned `url` or `text`, but never both at the same time.

Reply:
As discussed, both versions may hide some caveats, but I also think it's a not a too complicated function, so I just added comments in the existing version.